Add tweet mode parameter to relevant functions

JTP3XP opened the pull request:
Adding the tweet_mode parameter to functions that return tweets allows for specifying the "extended" tweet mode, which is how Twitter delivers the new longer tweets without truncating them. It will also allow for specifying "compat" mode when extended becomes the default.

zntfdr commented:
Surely this is the best way to add support to the ``extended`` mode without breaking legacy code. 
If @meteochu agrees, it looks good to me! 💯

meteochu commented:
Like I said, I still prefer using an enum approach, so that the implementation on the Swift side would feel more native. What we can do is having...

```swift
enum TweetMode {
    case default
    case extended
    case other(String) // in case twitter adds more in the future...

    var stringValue: String? {
        switch self {
        case .default: 
            return nil
        case .extended: 
            return "extended"
        case .other(let string):
            return string
        }
    }
}

// ...then your implementation can simply change to
parameters["tweet_mode"] ??= tweetMode.stringValue
```

This allows for the same flexibility while making it native.

JTP3XP commented:
Hey @meteochu, I think I see what you mean now. Xcode didn't like `default` being used like that, so I changed it to `defaultValue`, and added `compat` as an option to provide a clean way to force the current default so nothing changes when Twitter makes `extended` the default. 

Hopefully this is more like what you were thinking. I do think it seems much better than making everyone sift through the Twitter docs to figure out how to use the `tweet_mode` parameter. Let me know if you have any more thoughts on this latest version!

meteochu commented on Sources/Swifter.swift:
```diff
@@ -56,6 +56,27 @@ public enum TwitterURL {
 
+// MARK: - Tweet Mode
+public enum TweetMode {
+    case defaultValue
```
You can do the following to allow the keyword `default` to be used instead of having a longer name.
```
case `default`
```

JTP3XP commented:
@meteochu, sounds good. I just made that change.

meteochu commented on Sources/SwifterFavorites.swift:
```diff
@@ -34,25 +34,27 @@ public extension Swifter {
     If you do not provide either a user_id or screen_name to this method, it will assume you are requesting on behalf of the authenticating user. Specify one or the other for best results.
     */
-    public func getRecentlyFavouritedTweets(count: Int? = nil, sinceID: String? = nil, maxID: String? = nil, success: SuccessHandler? = nil, failure: FailureHandler? = nil) {
+    public func getRecentlyFavouritedTweets(count: Int? = nil, sinceID: String? = nil, maxID: String? = nil, tweetMode: TweetMode? = TweetMode.default, success: SuccessHandler? = nil, failure: FailureHandler? = nil) {
```
`tweetMode` should be a non-optional field as `default` has a `nil` raw value. This applies to everywhere else in the code.

JTP3XP commented:
@meteochu, that makes sense. No need for two different ways to specify a nil value - I made that change throughout the code.

zachrwolfe commented:
Would love to see this get merged; using it currently on my dev branch! Two nitpicks though:

First of all, IMO intent might be clearer if `tweetMode` were an optional field, the `default` case were removed, and `stringValue` were made non-optional. Passing nil more obviously communicates that you're not sending that parameter to Twitter than `.default`, which could be interpreted to mean either sending nothing or sending the text "default" until you look at the source code.

Second, all REST endpoint methods that return tweet objects should receive the tweet mode parameter, such as `postTweet`, `retweetTweet`, `unretweetTweet`, `destroyTweet`, etc.

meteochu commented:
By Swift styling design, `.default` is almost always a reserved keyword in an enum that refers to a "default" case. This will also allow for future-proofing if and when the default changes a couple years from now. I do agree that the remaining endpoints should include the tweet mode parameter; I will review this PR again once that is done.

zachrwolfe commented:
Fair enough!

JTP3XP commented:
I made another pass through the methods and hopefully got them all this time - I think the 2 favorite methods were the only outstanding methods @zachrwolfe did not mention.

meteochu commented:
LGTM, thank you for your contributions. This PR will now be merged.
